shpool/tests/attach.rs

```diff
@@ -1232,44 +1272,141 @@ fn motd_pager() -> anyhow::Result<()> {
         )
         .context("starting daemon proc")?;
 
-        // We need to manually spawn our attach proc because
-        // the motd gets printed immediately, so we can't always
-        // attach a line matcher in time.
-        let mut child = Command::new(support::shpool_bin()?)
-            .stdout(Stdio::piped())
-            .stderr(Stdio::piped())
-            .arg("--socket")
-            .arg(&daemon_proc.socket_path)
-            .arg("--config-file")
-            .arg(config_file)
-            .arg("attach")
-            .arg("sh1")
-            .spawn()
-            .context("spawning attach process")?;
+        let stdout_str = snapshot_attach_output(
+            &daemon_proc,
+            &config_file,
+            time::Duration::from_millis(500),
+            "sh1",
+        )?;
+        // Scan for a fragment of less output.
+        let stdout_file_re = Regex::new(".*\\(END\\).*")?;
+        assert!(stdout_file_re.is_match(&stdout_str));
 
-        // The attach shell should be spawned and have read the
-        // initial prompt after half a second.
-        std::thread::sleep(time::Duration::from_millis(500));
-        child.kill().context("killing child")?;
+        Ok(())
+    })
+}
 
-        let mut stderr = child.stderr.take().context("missing stderr")?;
-        let mut stderr_str = String::from("");
-        stderr.read_to_string(&mut stderr_str).context("slurping stderr")?;
-        assert!(stderr_str.is_empty());
+#[test]
+#[timeout(30000)]
+fn motd_debounced_pager_debounces() -> anyhow::Result<()> {
+    support::dump_err(|| {
+        // set up the config
+        let tmp_dir = tempfile::TempDir::with_prefix("shpool-test-config")?;
+        let tmp_dir_path = if env::var("SHPOOL_LEAVE_TEST_LOGS").is_ok() {
+            // leave the tmp files around for later inspection if we have been asked
+            // to leave the logs in place.
+            tmp_dir.into_path()
+        } else {
+            PathBuf::from(tmp_dir.path())
+        };
+        eprintln!("building config in {:?}", tmp_dir_path);
+        let motd_file = tmp_dir_path.join("motd.txt");
+        {
+            let mut f = fs::File::create(&motd_file)?;
+            f.write_all("MOTD_MSG\n".as_bytes())?;
+        }
+        let config_tmpl =
+            fs::read_to_string(support::testdata_file("motd_pager_1d_debounce.toml.tmpl"))?;
+        let config_contents = config_tmpl.replace("TMP_MOTD_MSG_FILE", motd_file.to_str().unwrap());
```

Comment:
Why not propagate the error with `?` similar to other errors? `motd_file` comes from `tmp_dir` which depends on system config could be non-valid utf-8. So this is not the case where we are sure `unwrap` won't fire.

Same in the other test.

Reply:
The error here doesn't cleanly convert to any anyhow::Error, so I would need to put a conversion routine to use ?. I decided to use unwrap because it is easier and this is just test code so a stack unwind is not a huge deal, just a slightly less convenient way of reporting an error.